File: sdk/runanywhere-react-native/src/Foundation/Configuration/ConfigurationService.ts
/**
 * ConfigurationService.ts
 *
 * Simple configuration service with fallback system
 *
 * Reference: sdk/runanywhere-swift/Sources/RunAnywhere/Data/Services/ConfigurationService.swift
 */

import { SDKLogger } from '../Logging/Logger/SDKLogger';

/**
 * Configuration data structure
 */
export interface ConfigurationData {
  readonly id: string;
  readonly source: string; // ConfigurationSource
  readonly routing?: any;
  readonly generation?: any;
  readonly storage?: any;
  readonly api?: any;
  readonly download?: any;
  readonly hardware?: any;
  readonly debugMode: boolean;
  readonly apiKey?: string;
  readonly allowUserOverride: boolean;
  readonly createdAt: Date;
  readonly updatedAt: Date;
}

/**
 * Simple configuration service with fallback system
 */
export class ConfigurationService {
  private logger: SDKLogger;
  private configRepository: any | null;
  private syncCoordinator: any | null;
  private currentConfig: ConfigurationData | null = null;

  constructor(
    configRepository?: any | null,
    syncCoordinator?: any | null
  ) {
    this.logger = new SDKLogger('ConfigurationService');
    this.configRepository = configRepository ?? null;
    this.syncCoordinator = syncCoordinator ?? null;
  }

  /**
   * Get current configuration
   */
  public getConfiguration(): ConfigurationData | null {
    return this.currentConfig;
  }

  /**
   * Load configuration on app launch
   */
  public async loadConfigurationOnLaunch(apiKey: string): Promise<ConfigurationData> {
    // Development mode: Skip remote fetch and use defaults
    if (!this.configRepository) {
      this.logger.info('Development mode: Using SDK defaults');
      const defaultConfig = this.getSDKDefaults(apiKey);
      this.currentConfig = defaultConfig;
      return defaultConfig;
    }

    // Step 1: Try to fetch remote configuration
    try {
      const remoteConfig = await this.configRepository.fetchRemoteConfiguration(apiKey);
      if (remoteConfig) {
        this.logger.info('Remote configuration loaded, saving to DB');
        await this.configRepository.save(remoteConfig);
        this.currentConfig = remoteConfig;
        return remoteConfig;
      }
    } catch (error) {
      // Continue to next step
    }

    // Step 2: Try to load from DB
    try {
      const dbConfig = await this.configRepository.fetch('default');
      if (dbConfig) {
        this.logger.info('Using DB configuration');
        this.currentConfig = dbConfig;
        return dbConfig;
      }
    } catch (error) {
      // Continue to next step
    }

    // Step 3: Try consumer configuration
    try {
      const consumerConfig = await this.configRepository.getConsumerConfiguration();
      if (consumerConfig) {
        this.logger.info('Using consumer configuration fallback');
        this.currentConfig = consumerConfig;
        return consumerConfig;
      }
    } catch (error) {
      // Continue to next step
    }

    // Step 4: Use SDK defaults
    this.logger.info('Using SDK default configuration');
    const defaultConfig = this.getSDKDefaults(apiKey);
    this.currentConfig = defaultConfig;
    return defaultConfig;
  }

  /**
   * Ensure configuration is loaded
   */
  public async ensureConfigurationLoaded(): Promise<void> {
    if (!this.currentConfig) {
      this.currentConfig = await this.loadConfigurationOnLaunch('');
    }
  }

  /**
   * Get SDK defaults
   */
  private getSDKDefaults(apiKey: string): ConfigurationData {
    return {
      id: 'default',
      source: 'defaults',
      debugMode: false,
      allowUserOverride: true,
      apiKey,
      createdAt: new Date(),
      updatedAt: new Date(),
    };
  }
}

File: sdk/runanywhere-swift/Sources/RunAnywhere/Infrastructure/Configuration/Services/ConfigurationService.swift
import Foundation

/// Simple configuration service with fallback system: DB → Consumer → SDK Defaults
public actor ConfigurationService: ConfigurationServiceProtocol {
    private let logger = SDKLogger(category: "ConfigurationService")
    private let configRepository: ConfigurationRepositoryImpl?
    private let syncCoordinator: SyncCoordinator?

    private var currentConfig: ConfigurationData?

    // MARK: - Initialization

    public init(configRepository: ConfigurationRepositoryImpl?, syncCoordinator: SyncCoordinator? = nil) {
        self.configRepository = configRepository
        self.syncCoordinator = syncCoordinator
        logger.info("ConfigurationService created\(configRepository == nil ? " (Development Mode)" : "")")
    }

    // MARK: - Public Methods

    public func getConfiguration() -> ConfigurationData? {
        return currentConfig
    }

    /// Load configuration on app launch with simple fallback: Remote → DB → Consumer → Defaults
    public func loadConfigurationOnLaunch(apiKey: String) async -> ConfigurationData {
        // Development mode: Skip remote fetch and use defaults
        guard let repository = configRepository else {
            logger.info("Development mode: Using SDK defaults")
            let defaultConfig = ConfigurationData.sdkDefaults(apiKey: apiKey)
            currentConfig = defaultConfig
            return defaultConfig
        }

        // Step 1: Try to fetch remote configuration
        if let remoteConfig = try? await repository.fetchRemoteConfiguration(apiKey: apiKey) {
            logger.info("Remote configuration loaded, saving to DB")
            // Save to DB and use as current config
            try? await repository.save(remoteConfig)
            currentConfig = remoteConfig
            return remoteConfig
        }

        // Step 2: Try to load from DB
        if let dbConfig = try? await repository.fetch(id: RegistryConstants.configurationId) {
            logger.info("Using DB configuration")
            currentConfig = dbConfig
            return dbConfig
        }

        // Step 3: Try consumer configuration
        if let consumerConfig = try? await repository.getConsumerConfiguration() {
            logger.info("Using consumer configuration fallback")
            currentConfig = consumerConfig
            return consumerConfig
        }

        // Step 4: Use SDK defaults
        logger.info("Using SDK default configuration")
        let defaultConfig = repository.getSDKDefaultConfiguration()
        currentConfig = defaultConfig
        return defaultConfig
    }

    /// Ensure configuration is loaded
    public func ensureConfigurationLoaded() async {
        if currentConfig == nil {
            currentConfig = await loadConfigurationOnLaunch(apiKey: "")
        }
    }

    /// Set consumer configuration override
    public func setConsumerConfiguration(_ config: ConfigurationData) async throws {
        guard let repository = configRepository else {
            logger.info("Development mode: Consumer configuration not persisted")
            currentConfig = config
            return
        }
        try await repository.setConsumerConfiguration(config)
        logger.info("Consumer configuration saved")
    }

    public func updateConfiguration(_ updates: (ConfigurationData) -> ConfigurationData) async {
        guard let config = currentConfig else {
            logger.warning("No configuration loaded")
            return
        }

        var updated = updates(config)

        // Development mode: Just update in memory
        guard let repository = configRepository else {
            currentConfig = updated
            logger.info("Development mode: Configuration updated in memory")
            return
        }

        do {
            // Mark as updated and save
            updated.markUpdated()
            try await repository.save(updated)

            // Trigger sync in background through coordinator
            if let syncCoordinator = syncCoordinator {
                Task {
                    try? await syncCoordinator.sync(repository)
                }
            }

            currentConfig = updated
            logger.info("Configuration updated, saved to DB and queued for sync")
        } catch {
            logger.error("Failed to save configuration: \(error)")
        }
    }

    public func syncToCloud() async throws {
        guard let repository = configRepository else {
            logger.info("Development mode: Sync skipped")
            return
        }

        // Sync through coordinator
        if let syncCoordinator = syncCoordinator {
            try await syncCoordinator.sync(repository)
        }
    }

    // MARK: - Factory Methods

    /// Create default configuration for development mode (no backend sync)
    /// - Parameter apiKey: The API key (can be empty for dev mode)
    /// - Returns: Default configuration data
    public func createDevelopmentModeConfig(apiKey: String) -> ConfigurationData {
        let config = ConfigurationData(
            id: "dev-\(UUID().uuidString)",
            apiKey: apiKey.isEmpty ? "dev-mode" : apiKey,
            source: .defaults
        )
        return config
    }
}
